src/thread_task.rs

```diff
@@ -142,180 +160,255 @@ fn reset_task_cur_async(ttd: &TaskThreadData, mut frame_idx: c_uint, n_frames: c
+                        || self.index(first).type_0 == TaskType::TileEntropy
+                );
+                assert!(self.index(first).type_0 == self.index(t).type_0);
+                assert!(self.index(t).sby == self.index(first).sby);
+                let t_tile_idx = self.index(first).tile_idx;
+                let p_tile_idx = self.index(t).tile_idx;
+                assert!(t_tile_idx != p_tile_idx);
+                if !(t_tile_idx > p_tile_idx) {
+                    self.insert_tasks_between(c, first, last, prev_t, t, cond_signal);
                     return;
                 }
-                // same task type
             }
+            // next:
+            prev_t = t;
+            t = self.index(t).next();
+        }
+        self.insert_tasks_between(c, first, last, prev_t, Rav1dTaskIndex::None, cond_signal);
+    }
+
+    fn push(&self, task: Rav1dTask) -> Rav1dTaskIndex {
+        let mut tasks = self.tasks.try_write().unwrap();
+        tasks.push(task);
+        // 1-based index into tasks, so we use length after pushing
+        Rav1dTaskIndex(NonZeroU32::new(tasks.len() as u32))
+    }
+
+    pub fn clear(&mut self) {
+        self.tasks.get_mut().clear();
+        self.pending_tasks.get_mut().clear();
+        self.pending_tasks_merge = AtomicBool::new(false);
+        self.head = Default::default();
+        self.cur_prev = Default::default();
+    }
+
+    pub fn remove(&self, t: Rav1dTaskIndex, prev_t: Rav1dTaskIndex) -> Option<Rav1dTask> {
+        let next_t = self.index(t).next();
+        if prev_t.is_some() {
+            self.index(prev_t)
+                .next
+                .compare_exchange(t, next_t, Ordering::SeqCst, Ordering::SeqCst)
+                .ok()?;
+        } else {
+            self.head
+                .compare_exchange(t, next_t, Ordering::SeqCst, Ordering::SeqCst)
+                .ok()?;
+        }
+        self.index(t).set_next(Rav1dTaskIndex::None);
+        Some(self.index(t).clone())
+    }
+
+    fn index<'a>(&'a self, index: Rav1dTaskIndex) -> impl Deref<Target = Rav1dTask> + 'a {
+        if let Some(index) = index.raw_index() {
+            RwLockReadGuard::map(self.tasks.try_read().unwrap(), |tasks| {
+                &tasks[index as usize]
+            })
+        } else {
+            panic!("Cannot index with None");
+        }
+    }
+
+    #[inline]
+    fn add_pending(&self, task: Rav1dTask) {
+        self.pending_tasks.lock().push(task);
+        self.pending_tasks_merge.store(true, Ordering::SeqCst);
+    }
+
+    #[inline]
+    fn merge_pending_frame(&self, c: &Rav1dContext) -> bool {
+        let merge = self.pending_tasks_merge.swap(false, Ordering::SeqCst);
+        if merge {
+            let mut pending_tasks = self.pending_tasks.lock();
+            let range = {
+                let mut tasks = self.tasks.try_write().unwrap();
+                if self.head.load(Ordering::Relaxed).is_none() {
```

Comment:
How come this is a relaxed load when all of the other `head` loads are `SeqCst`?